mne/beamformer/_compute_beamformer.py

```diff
@@ -253,96 +228,94 @@ def _compute_beamformer(G, Cm, reg, n_orient, weight_norm, pick_ori,
+        bf_numer = np.matmul(Gk.transpose(0, 2, 1), Cm_inv[np.newaxis])
+        bf_denom = np.matmul(bf_numer, Gk)
+        return bf_numer, bf_denom
+
+    #
+    # 2. Reorient lead field in direction of max power or normal
+    #
+    if pick_ori == 'max-power':
+        assert n_orient == 3
+        _, bf_denom = _compute_bf_terms(Gk, Cm_inv)
+        if weight_norm is None:
+            ori_numer = np.eye(n_orient)[np.newaxis]
+            ori_denom = bf_denom
         else:
-            # Rescale each filter separately
-            noise_norm = np.linalg.norm(W, axis=1, keepdims=True)
+            assert weight_norm in ['unit-noise-gain', 'nai']
+            # compute power, cf Sekihara & Nagarajan 2008, eq. 4.47
+            ori_numer = bf_denom
+            ori_denom = np.matmul(
+                np.matmul(Gk.transpose(0, 2, 1),
+                          Cm_inv.dot(Cm_inv)[np.newaxis]), Gk)
+        ori_denom_inv = _sym_inv_sm(ori_denom, reduce_rank, inversion, sk)
+        ori_pick = np.matmul(ori_denom_inv, ori_numer)
+        assert ori_pick.shape == (n_sources, n_orient, n_orient)
+
+        # pick eigenvector that corresponds to maximum eigenvalue:
+        eig_vals, eig_vecs = np.linalg.eig(ori_pick.real)  # not Hermitian!
+        # sort eigenvectors by eigenvalues for picking:
+        order = np.argsort(np.abs(eig_vals), axis=-1)
+        # eig_vals = np.take_along_axis(eig_vals, order, axis=-1)
+        max_power_ori = np.take_along_axis(
+            eig_vecs, order[:, np.newaxis, -1:], axis=-1)[:, :, 0]
+        assert max_power_ori.shape == (n_sources, n_orient)
+
+        # set the (otherwise arbitrary) sign to match the normal
+        signs = np.sign(np.sum(max_power_ori * nn, axis=1, keepdims=True))
+        max_power_ori *= signs
+
+        # Compute the lead field for the optimal orientation,
+        # and adjust numer/denom
+        Gk = np.matmul(Gk, max_power_ori[..., np.newaxis])
+        n_orient = 1
+    else:
+        max_power_ori = None
+        if pick_ori == 'normal':
+            Gk = Gk[..., 2:3]
+            n_orient = 1
+
+    #
+    # 3. Compute numerator and denominator of beamformer formula (unit-gain)
+    #
+
+    bf_numer, bf_denom = _compute_bf_terms(Gk, Cm_inv)
+    assert bf_denom.shape == (n_sources,) + (n_orient,) * 2
+    assert bf_numer.shape == (n_sources, n_orient, n_channels)
+    del Gk  # lead field has been adjusted and should not be used anymore
+
+    #
+    # 4. Invert the denominator
+    #
+
+    bf_denom_inv = _sym_inv_sm(bf_denom, reduce_rank, inversion, sk)
+    assert bf_denom_inv.shape == (n_sources, n_orient, n_orient)
+    W = np.matmul(bf_denom_inv, bf_numer)
+    assert W.shape == (n_sources, n_orient, n_channels)
+    del bf_numer, bf_denom, bf_denom_inv, sk
+
+    #
+    # 4. Re-scale filter weights according to the selected weight_norm
+    #
+
+    # Weight normalization is done by computing:
+    # W_ung = inv(sqrt(W_ug @ W_ug.T)) @ W_ug
```

Comment:
What will matter here is what sizes these things actually are. In other words, what is the size of W_ug, is it for a single ori `(1, n_channels)` or all three oris `(3, n_channels)`, and hence whether:

1. the `sqrt` is a plain square root
2. a matrix square root (only possible in the 3x3 case), or even
3. square root of the diagonal entries of the matrix only (unlikely but possible)

Currently we do a fourth thing, which is just to take the norm over the `(3, n_channels)` array, square root this singleton, and invert it, which doesn't act like any of the three above so is probably wrong.

Note that this only matters when `pick_ori='vector'` (or None) because if `pick_ori='normal'` or `pick_ori='max_power'`, we are in the 1-direction-per-source / `n_orient=1` case by this point in the code now.

Reply:
For now I've just implemented the direct formula with a Frobenius norm for each source point as this is both rotation invariant and had the lowest bias in the tests, but again I don't know if it's correct. I also implemented the non-rotation-invariant one below based on the formula you gave me @britta-wstnr, you can try it by uncommenting it (and commenting out the "direct way" code) if you want

Reply:
So far, the case where the matrix can be 3 x 3 has been wrong, I believe (vector case). I don't think it is a simple matrix square root in that case. 
I will check the different versions and re-check the formulas.